Personal/arnav/cancel enumeration fix scandir err

Strikerzee opened the pull request.

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -693,13 +697,11 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromTokenStore(ctx context.Context) (
 }
 
-// GetNewTokenFromMSI gets token from Azure Instance Metadata Service identity endpoint.
-// For details, please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
```
do write a function comment explaining what this function does. Highlight the fact that it sends a token request to the IMDS endpoint passed by the caller, which will be different for Azure and Arc VMs.

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,6 +722,34 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+	return req, resp, err
+}
+
+// GetNewTokenFromMSI gets token from Azure Instance Metadata Service identity endpoint.
```
update the comment to explain how it handles both Azure and Arc VMs by first trying the Azure IMDS endpoint and if that fails it tries the Arc IMDS endpoint.
Also update L730 to add Arc doc link too
https://docs.microsoft.com/en-us/azure/azure-arc/servers/managed-identity-authentication

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,6 +722,34 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+// For details, please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
+func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.Token, error) {
+	req, resp, err := credInfo.ProcessIMDSRequest(MSIEndpointAzureVM, ResourceAzureVM, IMDSAPIVersionAzureVM, ctx)
+	if err != nil {
```
today the msiTokenHTTPClient timeout is configured for 30 secs. We should discuss with Ze if we can reduce it to 5 secs. 5 secs should be enough as IMDS endpoint is local to the machine.
Without that, if some router is configured to not return "ICMP unreachable" then we will take 30 secs to timeout and fallback to ARC.

Strikerzee replied:
Changed to 10 seconds

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,6 +722,34 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+
+		challengeTokenPath := strings.Split(resp.Header["Www-Authenticate"][0], "=")[1]
+		// Open the file.
+		challengeTokenFile, _ := os.Open(challengeTokenPath)
```
we should collect err return from os.Open() and emit appropriate errors. For "not found" we should have a Fatal error saying Arc configuration looks improper, for "permission denied" we should warn and suggest that user should belong to himds group. We need to fist check how it behaves on windows.

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,6 +722,34 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+		// Open the file.
+		challengeTokenFile, _ := os.Open(challengeTokenPath)
+		// Create a new Scanner for the file.
+		scanner := bufio.NewScanner(challengeTokenFile)
```
bufio.NewReader() followed by ReadString() seems right for the job, since there's only one line and we don't really need a Scanner. Any specific reason you chose Scanner?

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,6 +722,34 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+	}
+
 	if err != nil {
 		return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm. (Error details: %v)", err)
```
shouldn't this error be Arc specific?
Azure specific error should already be emitted at L738

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+}
+
+// correctByteSlice corrects the value of JSON field "not_before" in the Byte slice from blank to a valid value and returns the corrected Byte slice.
+func correctByteSlice(bytes []byte) []byte {
```
nit. Let's call this fixupTokenJson() which makes the intent clearer.
And the comment above this should explain why we need this. I would put something like this.

// Dated 15th Sep 2021.
// Token JSON returned by ARC currently does not set a valid integral value for "not_before". 
// Fix will be available in future ARC versions, till then use the current time to fixup the byte slice before we pass it to json.Unmarshal().
// If the token JSON already has "not_before" correctly set, this will be a no-op.

Also, this should check if not_before has a null string and only then it should do the fixup, o/w it should be a no-op so that we can handle both cases where not_before is correctly set or not set.

Can we set it to "current time in epoch - 5 secs" just in case someone down the line actually interprets not_before and doesn't like an incorrect value.

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+	separatorString := "\"not_before\":\""
+	stringSlice := strings.Split(byteSliceToString, separatorString)
+	correctedString := stringSlice[0] + separatorString + "0" + stringSlice[1]
+	correctedByteSlice := []byte(correctedString)
```
nit,
return []byte(correctedString)

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+		// Try Arc VM
+		req, resp, err = credInfo.queryIMDS(MSIEndpointArcVM, Resource, IMDSAPIVersionArcVM, ctx)
+		if err != nil {
+			return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm: %v", err)
```
Please with P capital

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+		challengeTokenPath := strings.Split(resp.Header["Www-Authenticate"][0], "=")[1]
+		// Open the file.
+		challengeTokenFile, fileErr := os.Open(challengeTokenPath)
+		if os.IsPermission(fileErr) {
```
didn't the doc say that this is getting deprecated and we should use the other slightly longer statement?

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+		} else if fileErr != nil {
+			return nil, fmt.Errorf("Error occurred while opening file %s: %v", challengeTokenPath, fileErr)
+		}
+		// Create a new Reader for the file.
```
you should add the following after a successful open

defer challengeTokenFile.Close()

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +723,57 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+		reader := bufio.NewReader(challengeTokenFile)
+		challengeToken, fileErr := reader.ReadString('\n')
+		if fileErr != nil && fileErr != io.EOF {
+			return nil, fmt.Errorf("Error occurred while reading file: %v", fileErr)
```
"Error occurred while opening file %s: %v"

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -742,6 +794,7 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
 	result := &adal.Token{}
 	if len(b) > 0 {
 		b = ByteSliceExtension{ByteSlice: b}.RemoveBOM()
+		b = correctByteSlice(b)
```
currently it'll not work for Azure VMs which already have the not_before set correctly.
Once you fix it to conditionally set not_before, it'll be fine.

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +724,66 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+// fixupTokenJson corrects the value of JSON field "not_before" in the Byte slice from blank to a valid value and returns the corrected Byte slice.
+func fixupTokenJson(bytes []byte) []byte {
+	byteSliceToString := string(bytes)
+	separatorString := "\"not_before\":\""
```
nit, use backtick to avoid backslashes to improve readability.
`"not_before":"`

linuxsmiths commented on common/oauthTokenManager.go:
```diff
@@ -720,8 +724,66 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+	separatorString := "\"not_before\":\""
+	stringSlice := strings.Split(byteSliceToString, separatorString)
+	notBeforeTime := ""
+	if stringSlice[1][0] == '"' {
```
to play safe, if not_before is fine, return original slice.. something like

func fixupTokenJson(bytes []byte) []byte {
	byteSliceToString := string(bytes)
	separatorString := `"not_before":"`
	stringSlice := strings.Split(byteSliceToString, separatorString)
	if stringSlice[1][0] != '"' {
	    // Valid value for not_before, return original slice.
	    return bytes
	}

	// Stick in "now-5secs" as not_before value and return the updated slice.
    notBeforeTimeInteger := uint64(time.Now().Unix() - 5)
    notBeforeTime := strconv.FormatUint(notBeforeTimeInteger, 10)
	correctedString := stringSlice[0] + separatorString + notBeforeTime + stringSlice[1]
	return []byte(correctedString)
}

shalinijoshi19 reviewed (changes requested):
Hey @Strikerzee will you be moving this over to the Azure/Azure-storage-azcopy repo instead of the fork?

nitin-deamon committed:
This patch for enabling calling app to cancel enumeration.

- Used appCtx in parallel-walk so that when app want to cancel the
enumeration it can do.
- It doesn't effect azcopy working.

nitin-deamon committed:
Handle scanning directory errors in traverser.

* Fail visibly when scanning fails for files and directories

* add in more error paths

* fix processOneDirectory changes and add more comments

Commit:
Replace context pointer with context and nil with context.TODO()

Commit:
Minor fix.

Commit:
Fix nil ptr dereference.
